[Merged by Bors] - Allow enabling secure mode with Kerberos

sbernauer opened the pull request:
# Description

Closes https://github.com/stackabletech/hdfs-operator/issues/178
Fixes https://github.com/stackabletech/hdfs-operator/issues/338

TODOs

- [x] Release new Hadoop image with openssl and Kerberos clients use in docs and tests
- [x] Release and use operator-rs change
- [x] Fix hardcoded `kinit nn/simple-hdfs-namenode-default.default.svc.cluster.local@CLUSTER.LOCAL -kt /stackable/kerberos/keytab` in entrypoints
- [x] Go through all hadoop settings and see if they can be improved
- [X] Test different realms
- [x] Discuss CRD change
- [x] Discuss how to expose this in Discovery CM -> During on-site 2023/05 we have decided to ship this feature without exposing it via discovery *for now*
- [x] Implement discovery
- [x] Tests
- [x] Docs
- [x] Let  @maltesander have a look how we can better include the init container in the code structure
- [x] Test long running cluster (maybe turn down ticket lifetime for that)

<!-- Commit message above. Everything below is not added to the message. Do not change this line! -->

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [x] (Integration-)Test cases added
- [x] Documentation added or updated
- [ ] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

nightkr committed:
WIP: rekerberize

see #154

nightkr committed:
Merge branch 'main' into spike/security2

nightkr committed:
Workingish but incredibly hacky Kerberized HDFS

sbernauer committed:
Merge branch 'main' into spike/security2

sbernauer committed:
Use SecretOperatorVolumeSourceBuilder to build kerberos volume

sbernauer committed:
Add kdc and SecretClass to example

sbernauer committed:
Move ssl-server.xml and ssl-client.xml into constant

sbernauer committed:
WIP

sbernauer committed:
Only add volumes conditionally

sbernauer committed:
Use actual SecretClasses instead of hardcoding

sbernauer committed:
Remove HADOOP_JAAS_DEBUG=true

sbernauer committed:
Use add instead of extend

sbernauer committed:
Only add volume mounts when needed

sbernauer committed:
Only write either http or https nn address into hdfs-site

sbernauer committed:
Only mount tls certs to main containers and fix other stuff

sbernauer committed:
Use dedicated init container to create tls bundles

sbernauer committed:
Move kerberos service name into separate fn

sbernauer committed:
Dynamically determine principal name

sbernauer committed:
Read KERBEROS_REALM from krb5.conf rather than hardcoding

sbernauer committed:
WIP checkpoint, not working. Tries to use fqdn, nn fails to connect to jn because of DNS problems

sbernauer committed:
Swtich back to using service names as principals. Works again, only the datanode wait-for-namenodes init containers fails because it's stupid any requires the dfs.namenode.kerberos.principal setting (and ignores the dfs.namenode.kerberos.principal.pattern) (hdfs will start anyway)

sbernauer committed:
Only use hdfs_name within principal. It works!

sbernauer committed:
Test with more replicas

sbernauer committed:
Switch back to use fqdn principals

sbernauer committed:
fix warnings

sbernauer committed:
Only specify nn principal in wait-for-namenodes when kerberos is enabled

sbernauer committed:
Use operator-rs main branch

sbernauer committed:
Use new image

soenkeliebau committed:
Changed namespace of keytab secret to default so the example is directly deployable and works.

sbernauer committed:
Specify concrete principal when connecting to NN from init container

sbernauer committed:
Add first basic test

sbernauer committed:
fix typo

sbernauer committed:
Increase test number

sbernauer committed:
Add hadoop.security.authentication=kerberos to discovery CM

sbernauer committed:
Add a teststep to access hdfs. Get it running by adding stuff to discovery CM

sbernauer committed:
linter

sbernauer committed:
Set hdfs log level to DEBUG in test

sbernauer committed:
Increase log level and double number of test runs

sbernauer committed:
Disable logging again

sbernauer committed:
Test PROD.MYCORP realm

sbernauer committed:
Fix misc stuff

sbernauer committed:
Refactor stuff into kerberos.rs

sbernauer committed:
Create create_tls_cert_bundle_init_container_and_volumes fn

sbernauer committed:
Revert "Create create_tls_cert_bundle_init_container_and_volumes fn"

This reverts commit 0d81588a94d41932f5c84593d60c5718f7498f7c.

nightkr committed:
Add AD tests

nightkr committed:
Don't hard-code realm when ktiniting

sbernauer committed:
Disable activeDirectory tests

sbernauer committed:
Improve comment

sbernauer committed:
Write hadoop.kerberos.keytab.login.autorenewal.enabled to config and discovery CM

sbernauer committed:
Remove uneeded volume in tests

sbernauer committed:
Disable node principals by default

sbernauer committed:
Add wire encryption setting

sbernauer committed:
Merge branch 'main' into spike/security2

sbernauer committed:
Adopt to new secret-op crd

nightkr commented on Cargo.toml:
```diff
@@ -4,6 +4,6 @@ members = [
+[patch."https://github.com/stackabletech/operator-rs.git"]
 # stackable-operator = { path = "/home/sbernauer/stackabletech/operator-rs" }
-# stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "feat/affinities" }
+stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "main" }
```
Remove once op-rs is released

sbernauer replied:
Removed

nightkr commented on deploy/helm/hdfs-operator/crds/crds.yaml:
```diff
@@ -34,6 +34,40 @@ spec:
+                          default: false
+                          description: Wether a principal including the Kubernetes node name should be requested. The principal could e.g. be `HTTP/my-k8s-worker-0.mycorp.lan`. This feature is disabled by default, as the resulting principals can already by existent e.g. in Active Directory which can cause problems.
+                          type: boolean
+                        tlsSecretClass:
```
I'm not sure, I could see an argument for creating a `security` supergroup since this isn't strictly about Kerberos... On the other hand, how useful is it outside of that context?

sbernauer replied:
So you suggest something like the following?
```
      spec:
        clusterConfig:
          security:
            kerberos:
              tlsSecretClass: tls
              kerberosSecretClass: kerberos-$NAMESPACE
              wireEncryption: {{ test_scenario['values']['wire-encryption'] }}
```
I would say all these settings are security-related, yes. But for other product we have spec.clusterconfig.{authentication,authorization}, without putting it below spec.clusterconfig.security. So I guess the current way is more konsitent, WDYT?

nightkr replied:
More like:

```yaml
      spec:
        clusterConfig:
          security:
            tlsSecretClass: tls
            kerberos:
              kerberosSecretClass: kerberos-$NAMESPACE
            wireEncryption: {{ test_scenario['values']['wire-encryption'] }}
```

Since TLS (for example) doesn't have much to do with Kerberos compared to any other hypothetical authn method. Then again, there's the question of how likely that is to happen anyway...

sbernauer replied:
Implemented in 4a2afb92659f448ee25e2290459a67e73afebe6c

sbernauer replied:
We have decided on new structure in Arch meeting

nightkr commented on deploy/helm/hdfs-operator/crds/crds.yaml:
```diff
@@ -34,6 +34,40 @@ spec:
+                          default: tls
+                          description: Name of the SecretClass providing the tls certificates for the WebUIs.
+                          type: string
+                        wireEncryption:
```
If you want to shoot yourself in the foot then `configOverrides` already exists. No need for us to encourage this insanity.

sbernauer replied:
Asked in Slack: https://stackable-workspace.slack.com/archives/C031A56R127/p1684218706958919

sbernauer replied:
Removed the whole wire-encryption thing

nightkr commented on docs/modules/hdfs/examples/getting_started/zk.yaml:
```diff
@@ -10,4 +10,4 @@ spec:
     roleGroups:
       default:
-        replicas: 1
+        replicas: 3
```
I'm assuming this is not intended to be merged?

sbernauer replied:
Not entirely sure tbh ^^ Let's remove it again

nightkr commented on rust/operator/src/container.rs:
```diff
@@ -483,42 +585,117 @@ impl ContainerConfig {
+    fn export_kerberos_real_env_var_command() -> String {
+        "export KERBEROS_REALM=$(grep -oP 'default_realm = \\K.*' /stackable/kerberos/krb5.conf)"
+            .to_string()
+    }
```
I don't _think_ this dance should be required, if you try to `kinit` without a realm it should guess the realm automatically.

sbernauer replied:
We do not only use this in `kinit`, but also for
`.add("hadoop.registry.kerberos.realm", "${env.KERBEROS_REALM}")`

nightkr replied:
Fair.. This still feels hacky compared to parsing the file properly, but it's not like krb5 ships a better tool for it, and "use Python (or whatever) to FFI-call `krb5_get_default_realm`" sucks too.

sbernauer replied:
Agreed

nightkr replied:
https://github.com/stackabletech/docker-images/issues/371

nightkr replied:
I think this is ok for now, until we solve that issue.

nightkr commented on rust/operator/src/container.rs:
```diff
@@ -623,6 +824,15 @@ impl ContainerConfig {
+        if hdfs.has_kerberos_enabled() {
+            volume_mounts.push(VolumeMountBuilder::new("kerberos", "/stackable/kerberos").build());
+        }
+        if hdfs.https_secret_class().is_some() {
```
```suggestion
        if hdfs.has_https_enabled() {
```

nightkr commented on rust/operator/src/container.rs:
```diff
@@ -724,8 +930,15 @@ impl ContainerConfig {
+
+                if hdfs.has_kerberos_enabled() {
+                    jvm_args.push(
+                        "-Djava.security.krb5.conf=/stackable/kerberos/krb5.conf".to_string(),
```
Do we want to factor this out, or respect `HADOOP_OPTS` somehow?

sbernauer replied:
I have to admit, I didn't fully get that.
Just for context, this is the reason we split HADOOP_OPTS and HDFS_NAMENODE_OPTS https://github.com/stackabletech/hdfs-operator/blob/06ca693c1205b2474fd24a3d4ab5a7a37b3d20cb/rust/operator/src/container.rs#L662-L695

nightkr replied:
Yes, but.. this will break existing users who already rely on setting a `HADOOP_OPTS` override, in the sense that their settings will no longer take effect, right? Unless I'm missing something here.

sbernauer replied:
If a users currently overrides HADOOP_OPTS this override will still take precedence, so it should not break any existing setup. The implications is that some parts might not work because the miss the `java.security.krb5.conf` setting *when enabling kerberos*.
But i think this is the usual risk when using overrides.

nightkr replied:
OK.

nightkr commented on rust/operator/src/container.rs:
```diff
@@ -724,8 +930,15 @@ impl ContainerConfig {
+                }
+
+                if let Some(Some(Some(memory_limit))) =
+                    resources.map(|r| r.limits.as_ref().map(|limits| limits.get("memory")))
```
Nit:

```suggestion
                if let Some(memory_limit) =
                    resources.and_then(|r| r.limits.as_ref()?.get("memory"))
```

nightkr commented on rust/operator/src/hdfs_controller.rs:
```diff
@@ -415,10 +428,69 @@ fn rolegroup_config_map(
                     .extend(config)
                     .build_as_xml();
             }
+            // PropertyNameKind::File(file_name) if file_name == HADOOP_POLICY_XML => {
```
Legacy/testing stuffs?

sbernauer replied:
We did run into the error mentioned in comments. I have no idea what I did to resolve them, but it seemed to have worked. Let's remove them, we can always get them from the history again

nightkr commented on rust/operator/src/kerberos.rs:
```diff
@@ -0,0 +1,139 @@
+                .add("hadoop.security.authorization", "true")
+                // Otherwise we fail with `java.io.IOException: No groups found for user nn`
+                // Default value is `dr.who=`, so we include that here
+                .add("hadoop.user.group.static.mapping.overrides", "dr.who=;nn=;nm=;jn=;")
```
Note to self that this is a placeholder until we have group mapping (https://github.com/stackabletech/opa-operator/pull/433)

sbernauer replied:
Closing, as this was a note to self

nightkr commented on tests/templates/kuttl/kerberos/02-create-kerberos-secretclass.yaml.j2:
```diff
@@ -0,0 +1,72 @@
+  # wkt /tmp/keytab
+  # EOF
+{% if test_scenario['values']['kerberos-realm'] == 'CLUSTER.LOCAL' %}
+  keytab: BQIAAABdAAEADUNMVVNURVIuTE9DQUwAGXN0YWNrYWJsZS1zZWNyZXQtb3BlcmF0b3IAAAABZAYWIgEAFAAgm8MCZ8B//XF1tH92GciD6/usWUNAmBTZnZQxLua2TkgAAAAB
```
Probably want to decide how to handle these before merging

sbernauer replied:
Do you see any room for improvement? I personally would be fine with merging as-is, but also happy about improvement ideas :see_no_evil:

nightkr replied:
I think the ideal would be to generate the keytabs on demand, but that also brings downsides.. bleh

sbernauer replied:
Yeah, it's tricky...

sbernauer replied:
I think we did our reasonable best in the PR, I think further improvements would be pretty complicated and could be carried out in secret-op tests first

sbernauer committed:
Merge remote-tracking branch 'origin/main' into spike/security2

sbernauer committed:
Fix merge mistake

sbernauer committed:
Bump tests to 23.4 and respect listenerClass

sbernauer committed:
Increase assert timeout

sbernauer committed:
Only run kerberos tests

sbernauer committed:
Merge branch 'main' into spike/security2

sbernauer committed:
Revert example replicas to 1

sbernauer committed:
Update rust/operator/src/container.rs

Co-authored-by: Natalie <teo@nullable.se>

sbernauer committed:
Update rust/operator/src/container.rs

Co-authored-by: Natalie <teo@nullable.se>

sbernauer committed:
Remove commented out code regarding HADOOP_POLICY_XML

sbernauer committed:
fix format

sbernauer committed:
Address Arch discussion feedback

sbernauer committed:
Add docs

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+= Security
+
+== Authentication
+Currently the only authentication mechanism support is Kerberos, which is disabled by default.
```
```suggestion
Currently the only supported authentication mechanism is Kerberos, which is disabled by default.
```

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+
+== Authentication
+Currently the only authentication mechanism support is Kerberos, which is disabled by default.
+For Kerberos to work a Kerberos server is needed, which the users needs to provide.
```
```suggestion
For Kerberos to work a Kerberos KDC is needed, which the users needs to provide.
```

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+Currently the only authentication mechanism support is Kerberos, which is disabled by default.
+For Kerberos to work a Kerberos server is needed, which the users needs to provide.
+The xref:home:secret-operator:secretclass.adoc#backend-kerberoskeytab[secret-operator documentation] states which kind of Kerberos servers are supported and how they can be configured.
+Our integration tests use a mit krb5 server which is installed on Kubernetes, but is not production ready, e.g. because no high availability is configured.
```
```suggestion
Our integration tests use a MIT krb5 server which is installed on Kubernetes. Our test setup is, however, not production ready. For example, it does not configure any high availability.
```

nightkr replied:
Honestly I'd almost strike this whole line entirely.

sbernauer replied:
striked :heavy_check_mark:

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+To configure HDFS to use Kerberos you first need to collect information about your Kerberos server, e.g. hostname and port.
+Additionally you need a service-user, which the secret-operator uses to create create principals for the HDFS services.
+
+=== 2. Create SecretClass
```
```suggestion
=== 2. Create Kerberos SecretClass
```

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+
+The `tlsSecretClass` is needed to request TLS certificates, used e.g. for the Web UIs.
+
+As an alternative you can
```
```suggestion

```

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+
+As an alternative you can
+
+=== 4. Verify Kerberos is used
```
```suggestion
=== 4. Verify that Kerberos is used
```

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+As an alternative you can
+
+=== 4. Verify Kerberos is used
+One option is to use `stackablectl services list --all-namespaces` to get the endpoint the HDFS namenodes are reachable.
```
```suggestion
Use `stackablectl services list --all-namespaces` to get the endpoints where the HDFS namenodes are reachable.
```

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+
+> Security is on.
+
+An alternative is to shell into the namenode and try to access the file system.
```
```suggestion
You can alsoshell into the namenode and try to access the file system.
```

sbernauer replied:
Changed to `You can also shell into the namenode and try to access the file system:`

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+An alternative is to shell into the namenode and try to access the file system.
+`kubectl exec -it hdfs-namenode-default-0 -c namenode -- bash -c 'bin/hdfs dfs -ls /'`
+
+You should get a `org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]`.
```
```suggestion
You should get the error message `org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]`.
```

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+You should get a `org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]`.
+
+=== 5. Access HDFS
+In case you want to access your HDFS it is recommended to start up a client Pod that connects to HDFS over shelling into the namenode.
```
```suggestion
In case you want to access your HDFS it is recommended to start up a client Pod that connects to HDFS, rather than shelling into the namenode.
```

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+|In addition to Authentication, it guarantees that a man-in-the-middle cannot tamper with messages exchanged between the client and the server.
+Sets `hadoop.rpc.protection` to `integrity`, `hadoop.data.transfer.protection` to `integrity` and `dfs.encrypt.data.transfer` to `false`.
+
+|Privacy
```
```suggestion
|Privacy (default)
```

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -0,0 +1,109 @@
+    kerberos:
+      # kerberosSecretClass: kerberos
+      wireEncryption: Privacy
+----
```
```suggestion
It is strongly recommended to leave `wireEncryption` in its default `Privacy` mode, since `Authentication` and `Integrity` leave the system vulnerable to man-in-the-middle and traffic sniffing attacks, respectively.
```

sbernauer replied:
I would prefer to document the wire-encryption setting. WDYT of

In case kerberos is enabled, the default value is `Privacy`, since `Authentication` and `Integrity` leave the system vulnerable to man-in-the-middle and traffic sniffing attacks, respectively.
The security measures come with a cost of a potentially degraded performance, thus wire encryption can be disabled, but is recommend to be enabled for security reasons.

You can specify the wire encryption mode to use as follows:
[...]

nightkr replied:
I definitely don't think it's appropriate to have language that could encourage users to touch it in a guide document like this.

sbernauer replied:
IMHO it's a totally legit use-case to enable Kerberos but don't use wire-encryption (I would say every hdfs clusters I have used so far had this setup (if Kerberos was used at all))

sbernauer replied:
Sorry, you changed your comment. Do you have an idea what can we add to make it more even more clear to not disable wire encryption?

nightkr replied:
> IMHO it's a totally legit use-case to enable Kerberos but don't use wire-encryption (I would say every hdfs clusters I have used so far had this setup (if Kerberos was used at all))

I think this is the crux of the disagreement though. What's the scenario where you do care about authentication but not about having it be useful for anything?

sbernauer replied:
let me cheekily copy Lars' words:

> Kerberos itself is a secure authentication protocol already so it is independent of wire encryption and there might be use-cases where people want to have best possible performance and don't care about whether anyone can read the data on the network (e.g. other protection mechanisms)

sbernauer replied:
Removed the whole wire-encryption thing

nightkr commented on tests/test-definition.yaml:
```diff
@@ -2,12 +2,12 @@
-      - 3.3.3-stackable23.4.0-rc3
-      - 3.3.4-stackable23.4.0-rc3
+      # - 3.2.2-stackable0.0.0-dev
+      # - 3.3.3-stackable0.0.0-dev
```
Are we dropping support for these?

sbernauer replied:
I only removed `3.3.3` in favor of `3.3.4` to keep the number of tests manageable

nightkr replied:
OK.

nightkr commented on tests/test-definition.yaml:
```diff
@@ -35,31 +40,33 @@ dimensions:
+  #     - zookeeper
+  #     - number-of-datanodes
+  #     - datanode-pvcs
+  #     - listener-class
```
Reenable before merging??

sbernauer replied:
reenabled

nightkr commented on tests/test-definition.yaml:
```diff
@@ -35,31 +40,33 @@ dimensions:
+  # - name: cluster-operation
+  #   dimensions:
+  #     - hadoop-latest
+  #     - zookeeper-latest
```
Reenable before merging?

sbernauer replied:
reenabled

sbernauer committed:
Update docs/modules/hdfs/pages/usage-guide/security.adoc

Co-authored-by: Natalie <teo@nullable.se>

sbernauer committed:
Update docs/modules/hdfs/pages/usage-guide/security.adoc

Co-authored-by: Natalie <teo@nullable.se>

sbernauer committed:
Update docs/modules/hdfs/pages/usage-guide/security.adoc

Co-authored-by: Natalie <teo@nullable.se>

sbernauer committed:
Apply suggestions from code review

Co-authored-by: Natalie <teo@nullable.se>

sbernauer committed:
Apply review comment

sbernauer committed:
Remove sentence

sbernauer committed:
Re-enable tests

sbernauer committed:
Rework CRD accoring to review feedback

sbernauer committed:
Rename kerberos.kerberosSecretClass to kerberos.secretClass

nightkr reviewed (commented):
LGTM, modulo the questions that are still open.

sbernauer committed:
Adress Arch meeting feedback

sbernauer committed:
charts

sbernauer commented:
@nightkr adopted the CRD to the Arch meeting decision. Removed the wire-encryption entirely.
Please have another look!

sbernauer commented:
Started test https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hdfs-operator-it-custom/72/
I will also spin up an HDFS and have a look at it tomorrow

sbernauer committed:
Re-enable all test cases

sbernauer committed:
Re-add wire encryption privacy settings

sbernauer commented:
Next try: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hdfs-operator-it-custom/73

sbernauer commented:
Tests are green again.
Long-running hdfs clusters are fine as well, but well have to wait a bit more
```
Default principal: nn/hdfs-namenode-default.kuttl-test-adjusted-oarfish.svc.cluster.local@PROD.MYCORP

Valid starting     Expires            Service principal
05/24/23 13:57:11  05/25/23 13:57:11  krbtgt/PROD.MYCORP@PROD.MYCORP
```

sbernauer commented:
All my namenodes are crashing with
`org.apache.hadoop.hdfs.server.namenode.EditLogInputException: Error replaying edit log at offset 0.  Expected transaction ID was 1`
`Caused by: org.apache.hadoop.hdfs.server.namenode.RedundantEditLogInputStream$PrematureEOFException: got premature end-of-file at txid 0; expected file to go up to 10 `

I guess the reason is that secret-op writes an annotation `restarter.stackable.tech/expires-at.423b4dce15bb8d736818cf7067bc2d91: "2023-05-26T12:06:22.264128554+00:00"` because we get a TLS cert from it with a lifetime of 24h.

I don't *think* this is a Kerberos problem, but we are propably running into https://github.com/stackabletech/hdfs-operator/issues/338.
This causes real problems, as all journalnodes will restart at the same time thus bricking the HDFS :(
And if this does not cause problems, https://github.com/stackabletech/hdfs-operator/issues/264 will propably.

CC @lfrancke for prioritization of https://github.com/stackabletech/hdfs-operator/issues/338

nightkr commented:
See also: https://github.com/stackabletech/secret-operator/issues/159. Arguably we should fix both that and #338.

sbernauer commented:
Just for the record: Maybe PodDisruptionBudgets could be a quick win. I think evictions (from commons-op) should respect that, but I'm only on my phone

sbernauer committed:
fix: Only add truststore settings when https is enabled

sbernauer commented:
Quick update: I'm able to reproduce the problem when execution `kubectl -n kuttl-test-pumped-maggot delete pod hdfs-journalnode-default-0 hdfs-journalnode-default-1 hdfs-journalnode-default-2`
Strangely if I do this with non-kerberized Cluster everything is fine :thinking: (This clusters even survive `kubectl -n kuttl-test-fitting-silkworm delete pod hdfs-zk-server-default-0 hdfs-zk-server-default-1 hdfs-zk-server-default-2 hdfs-journalnode-default-0 hdfs-journalnode-default-1 hdfs-journalnode-default-2 hdfs-namenode-default-0 hdfs-namenode-default-1 hdfs-datanode-default-0 hdfs-datanode-default-1`)

sbernauer commented on rust/operator/src/container.rs:
```diff
@@ -391,26 +482,31 @@ impl ContainerConfig {
 
                     set -e
-                    if [ ! -f "{NAMENODE_ROOT_DATA_DIR}/current/VERSION" ]
+                    if ! ls {NAMENODE_ROOT_DATA_DIR}/current/fsimage_*
```
Maybe this causes the JN problems. @nightkr *only if you have the time and inclination*: This is from your spike, what was the intention of changing this?
Will give the old code a test and see if the problem persists

sbernauer replied:
Ok, didn't change anything :/

soenkeliebau replied:
Would probably cause breakage when upgrading the operator later on though, similar to when we renamed datanode paths.

sbernauer replied:
Would it? This only changes the command in the init container

sbernauer replied:
This change is actually correct and needed. Feel free to reopen if desired!

sbernauer committed:
test: Switch image to nightly

sbernauer committed:
Also add ssl.server.truststore.location setting

sbernauer committed:
Add comment

sbernauer committed:
Remove redundant set -e

sbernauer committed:
We only support Kerberos for HDFS >= 3.3.x

sbernauer committed:
Add chaos monkey to smoke test

sbernauer committed:
Merge remote-tracking branch 'origin/main' into spike/security2

sbernauer commented:
Started test https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hdfs-operator-it-custom/76/

sbernauer committed:
docs

sbernauer committed:
Use constants consistently

sbernauer committed:
Simplify

sbernauer committed:
Minor improvements

sbernauer committed:
Fix kerberos hdfs version check

sbernauer committed:
cleanup

sbernauer committed:
intendation

sbernauer committed:
Sett hadoop.rpc.protection to privacy

sbernauer committed:
Check exit code of zk format

sbernauer committed:
Reduce test duration by restricting chaosmonkey in smoke test

sbernauer committed:
Force delte pods

nightkr commented on docs/modules/hdfs/pages/usage-guide/security.adoc:
```diff
@@ -43,7 +45,7 @@ The important part is
 
 You can also shell into the namenode and try to access the file system:
-`kubectl exec -it hdfs-namenode-default-0 -c namenode -- bash -c 'bin/hdfs dfs -ls /'`
+`kubectl exec -it hdfs-namenode-default-0 -c namenode -- bash -c 'kdestroy && bin/hdfs dfs -ls /'`
```
This paragraph doesn't really have the same oomph if we're kdestroying. Maybe we should change it to a `kubectl run` of some kind with a temporary pod instead?

sbernauer replied:
Agreed. Would it be ok to get rid of this command entirely and rely on the Webinterface?

nightkr replied:
I like demonstrating it with the CLI too, honestly. Maybe that's something to add to the stackablectl wishlist at some point...

nightkr replied:
Do we still need to `kdestroy` now that we're back to CC-per-container? Actually, I think we could just override the CC path for that specific command...

sbernauer replied:
Current PR state has no kdestroy any more :+1:

nightkr commented on rust/operator/src/container.rs:
```diff
@@ -161,16 +157,22 @@ impl ContainerConfig {
-            object_name,
         )?);
 
+        // We need to share `/tmp` between all containers, e.g. for Kerberos ticket cache
```
We shouldn't need to share the ticket cache as long as every container has access to the keytab itself (and container B relying on container A having populated the cache already is a great way to run into bugs when those tickets start expiring, for example).

sbernauer replied:
Implemented in 01d4a5a895bf8bddb4f18be66968e13af3e9e570

nightkr commented on rust/operator/src/container.rs:
```diff
@@ -199,13 +203,40 @@ impl ContainerConfig {
+                            export KRB5_CONFIG=/stackable/kerberos/krb5.conf
+                            {export_kerberos_real_env_var_command}
+                            echo "Getting ticket for {principal} from /stackable/kerberos/keytab"
+                            kinit "{principal}" -kt /stackable/kerberos/keytab
```
At most this should only be relied on by other init containers, since these tickets expire (as mentioned above)...

sbernauer replied:
Implemented in 01d4a5a895bf8bddb4f18be66968e13af3e9e570

nightkr commented on rust/operator/src/container.rs:
```diff
@@ -199,13 +203,40 @@ impl ContainerConfig {
                     .build(),
             );
 
+            pb.add_volume(
+                VolumeBuilder::new("kerberos")
+                    .ephemeral(
+                        SecretOperatorVolumeSourceBuilder::new(
+                            &authentication_config.kerberos.secret_class,
+                        )
+                        .with_service_scope(hdfs.name_any())
+                        .with_kerberos_service_name(role.kerberos_service_name())
+                        .with_kerberos_service_name("HTTP")
+                        .build(),
+                    )
+                    .build(),
+            );
```
:+1:

nightkr commented on rust/operator/src/kerberos.rs:
```diff
@@ -1,9 +1,21 @@
+};
+
+pub fn check_if_supported(resolved_product_image: &ResolvedProductImage) -> Result<(), Error> {
+    if resolved_product_image.product_version.starts_with("3.2.") {
```
It feels hacky but I guess it's simple enough and works... Especially since we can drop it when we drop 3.2.x support.

sbernauer replied:
Exactly my thoughts :+1:

nightkr commented on rust/operator/src/kerberos.rs:
```diff
@@ -22,80 +35,174 @@ impl HdfsSiteConfigBuilder {
+    }
 }
+
+// IMPORTANT: We only support Kerberos for HDFS >= 3.3.x
```
Can we move this section up into the `check_if_supported` fn? Maybe with the full stack trace snipped out or moved into a gist/separate file.

sbernauer replied:
I would prefer to keep the full stacktrace, just to be sure and have be prepared for customer questions.
Questions is where should I put it? A file `rust/operator/src/kerberos_hdfs_3.2_stacktrace.txt`?

nightkr replied:
Sure.

sbernauer replied:
277c5503ffae6962acd3e55020dff9b0edc8194a

nightkr commented on tests/templates/kuttl/kerberos/21-unleash-the-chaosmonkey.yaml.j2:
```diff
@@ -0,0 +1,49 @@
+# Tribute to https://github.com/Netflix/chaosmonkey
+
+# We need to force-delete the Pods, because IONOS is sometimes unable to delete the pod (it's stuck in Terminating for > 20 minutes)
```
That's worrying... do we have access to IONOS' kubelet logs? Is this an IONOS-only issue?

sbernauer replied:
Have to admit I'm not sure if we have access to the kubelet logs. Test passed on Azure however

sbernauer replied:
Update: Support recommended using https://github.com/kvaps/kubectl-node-shell, which worked.
Sadly it only logs stuff during startup and than remains silent.
```
journalctl -u kubelet --since "48 hour ago"
-- Logs begin at Sat 2023-06-03 19:51:35 UTC, end at Tue 2023-06-06 12:01:05 UTC. --
-- No entries --
```
Anyway, even with force deletion tests have random timeouts on IONOS after we added the chaosmonkey. Everything is really slow

nightkr commented on tests/templates/kuttl/kerberos/21-unleash-the-chaosmonkey.yaml.j2:
```diff
@@ -0,0 +1,49 @@
+# Tribute to https://github.com/Netflix/chaosmonkey
```
Maybe add a link to #338 too, as a motivation for why this exists?

sbernauer replied:
Added in bfaafa4ffe28c96d9eb69c302b799f621bff64df

nightkr reviewed (commented):
Mostly LGTM in theory, a few nits and haven't tested it yet.

sbernauer committed:
Link to journalnodes bug ticket

sbernauer committed:
Don't share /tmp ticket cache between containers

sbernauer committed:
Removed uneeded -D dfs.namenode.kerberos.principal=$PRINCIPAL

This is not needed any more after changing the princial names to not contain _HOST

sbernauer committed:
Move stacktrace into dedicated file

sbernauer committed:
Only kinit in init(!) contains that need a ticket

sbernauer committed:
fix: Set KERBEROS_REALM in every container, not only the ones with kinit

sbernauer commented:
Should be ready to be merged :+1:

sbernauer committed:
fix: Only kinit when Kerberos is enabled ;)

nightkr reviewed (approved):
Waiting for integration tests to complete, but LGTM otherwise.

We'll probably want to note this in the changelog though?

nightkr commented:
Tests passed.

sbernauer committed:
changelog

sbernauer commented:
All that and I forgot the CHANGELOG :D
Thanks for your review and work + help on Kerberos overall!

I have 4 testclusters running since this morning, let's wait some time with merging to see if they survive

sbernauer commented:
All 4 clusters are still healthy, merging this

sbernauer commented:
bors r+
